src/crystal/system/unix/user.cr

```diff
@@ -36,14 +39,17 @@ module Crystal::System::User
     initial_buf = uninitialized UInt8[1024]
     buf = initial_buf.to_slice
 
-    ret = LibC.getpwuid_r(id, pwd_pointer, buf, buf.size, pointerof(pwd_pointer))
-    while ret == LibC::ERANGE && buf.size < GETPW_R_SIZE_MAX
-      buf = Bytes.new(buf.size * 2)
-      ret = LibC.getpwuid_r(id, pwd_pointer, buf, buf.size, pointerof(pwd_pointer))
+    while (ret = LibC.getpwuid_r(id, pwd_pointer, buf, buf.size, pointerof(pwd_pointer))) != 0
+      case ret
+      when LibC::ENOENT, LibC::ESRCH, LibC::EBADF, LibC::EPERM
+        return nil
+      when LibC::ERANGE
+        raise RuntimeError.from_errno("getpwuid_r") if buf.size >= GETPW_R_SIZE_MAX
+        buf = Bytes.new(buf.size * 2)
+      else
+        raise RuntimeError.from_errno("getpwuid_r")
+      end
```

Comment:
The code for all four methods looks very similar now, except for a few differences like what's the function to call and the name to use for `from_errno`. Could we extract them to a separate method that yields? That way all that logic will be in a single place. Or maybe not a method that yields, I'm sure there's a more DRY way to do this.

Reply:
I created a macro to wrap the functions. 
I also tried putting it in a method that yields. But that was too complicated.